### PR TITLE
Multi-attach feature for hotpot.

### PR DESCRIPTION
**Pull request by jackhaibo**
**What this PR does / why we need it**:
Multi-attach for sushi, attach volume more than once.
Please see design spec for details.
opensds/design-specs#26

### CODE REVIEW

**Comment by leonwanghui** [comment]
@wisererik @xing-yang @xiangrumei Please take a review on this PR.

File: csi/server/plugin/opensds/node.go [hunk 1]
```diff
@@ -118,53 +126,61 @@ func getVolumeAndAttachmentByVolumeId(volId string) (*model.VolumeSpec, *model.V
 		}
 	}
 
+	if attachment == nil {
+		msg := fmt.Sprintf("attachment is not exist")
```
Comment (wisererik) [inline csi/server/plugin/opensds/node.go]: attachment does not exist

File: csi/server/plugin/opensds/node.go [hunk 2]
```diff
@@ -118,53 +126,61 @@ func getVolumeAndAttachmentByVolumeId(volId string) (*model.VolumeSpec, *model.V
...
+		attachment.Metadata[key] = strings.Join(paths, ";")
 	}
 
 	if KStagingTargetPath == key {
 		volConnector := connector.NewConnector(attachment.DriverVolumeType)
 
 		if volConnector == nil {
-			return status.Error(codes.FailedPrecondition, fmt.Sprintf("Unsupport driverVolumeType: %s", attachment.DriverVolumeType))
+			msg := fmt.Sprintf("unsupport driverVolumeType: %s", attachment.DriverVolumeType)
+			glog.Error(msg)
```
Comment (wisererik) [inline csi/server/plugin/opensds/node.go]: unsupported driver volume type

File: csi/server/plugin/opensds/node.go [hunk 3]
```diff
@@ -203,15 +219,20 @@ func (p *Plugin) NodeStageVolume(
 	*csi.NodeStageVolumeResponse, error) {
 
 	// Check REQUIRED field
-	glog.V(5).Info("start to NodeStageVolume, Volume_id: " + req.VolumeId + ", staging_target_path: " + req.StagingTargetPath)
+	glog.V(5).Info("start to NodeStageVolume, Volume_id: " + req.VolumeId +
+		", staging_target_path: " + req.StagingTargetPath)
```
Comment (wisererik) [inline csi/server/plugin/opensds/node.go]: start to node stage volume, volume_id

File: csi/server/plugin/opensds/node.go [hunk 4]
```diff
@@ -264,7 +285,9 @@ func (p *Plugin) NodeStageVolume(
 	glog.V(5).Infof("VolumeCapability Mount=%+v, Block=%+v\n", mnt, block)
 
 	if nil != mnt && nil != block {
-		return nil, status.Error(codes.InvalidArgument, "volumeMode cannot be both Block and Filesystem")
+		msg := "volumeMode cannot be both Block and Filesystem"
+		glog.Error(msg)
```
Comment (wisererik) [inline csi/server/plugin/opensds/node.go]: same here

File: csi/server/plugin/opensds/node.go [hunk 5]
```diff
@@ -442,20 +484,23 @@ func (p *Plugin) NodePublishVolume(
...
```
Comment (wisererik) [inline csi/server/plugin/opensds/node.go]: same here

File: csi/server/plugin/opensds/node.go [hunk 6]
```diff
@@ -584,21 +635,21 @@ func (p *Plugin) NodeGetInfo(
 	for _, volDriverType := range volDriverTypes {
 		volDriver := connector.NewConnector(volDriverType)
 		if volDriver == nil {
-			glog.Errorf("Unsupport volDriver: %s", volDriverType)
+			glog.Errorf("unsupport volDriver: %s", volDriverType)
```
Comment (wisererik) [inline csi/server/plugin/opensds/node.go]: unsupported volume driver

Comment (jackhaibo) [reply csi/server/plugin/opensds/node.go]: Done.

File: csi/server/plugin/opensds/node.go (hunk 5 above)
Comment (jackhaibo) [reply csi/server/plugin/opensds/node.go]: Done.

File: csi/server/plugin/opensds/node.go (hunk 4 above)
Comment (jackhaibo) [reply csi/server/plugin/opensds/node.go]: Done.

File: csi/server/plugin/opensds/node.go (hunk 3 above)
Comment (jackhaibo) [reply csi/server/plugin/opensds/node.go]: Done.

File: csi/server/plugin/opensds/node.go (hunk 2 above)
Comment (jackhaibo) [reply csi/server/plugin/opensds/node.go]: Done.

File: csi/server/plugin/opensds/node.go (hunk 1 above)
Comment (jackhaibo) [reply csi/server/plugin/opensds/node.go]: Done

File: csi/server/plugin/opensds/controller.go [hunk 7]
```diff
@@ -384,90 +378,100 @@ func isStringMapEqual(metadataA, metadataB map[string]string) bool {
...
+		msg := "controllerPublishVolume: Volume ID must be provided"
+		glog.Error(msg)
 		return nil, status.Error(codes.InvalidArgument, msg)
 	}
 
 	if req.NodeId == "" {
-		msg := "ControllerPublishVolume Node ID must be provided"
-		glog.Info(msg)
+		msg := "controllerPublishVolume: Node ID must be provided"
+		glog.Error(msg)
```
[the remark this answers is not in the document]
Comment (jackhaibo) [reply csi/server/plugin/opensds/controller.go]: Done.

File: csi/server/plugin/opensds/controller.go [hunk 8]
```diff
@@ -384,90 +378,100 @@ func isStringMapEqual(metadataA, metadataB map[string]string) bool {
...
 	req *csi.ControllerPublishVolumeRequest) (
 	*csi.ControllerPublishVolumeResponse, error) {
 
 	glog.V(5).Info("start to ControllerPublishVolume")
 	defer glog.V(5).Info("end to ControllerPublishVolume")
 
 	if req.VolumeId == "" {
-		msg := "ControllerPublishVolume Volume ID must be provided"
-		glog.Info(msg)
+		msg := "controllerPublishVolume: Volume ID must be provided"
```
[the remark this answers is not in the document]
Comment (jackhaibo) [reply csi/server/plugin/opensds/controller.go]: Done.

File: csi/server/plugin/opensds/controller.go [hunk 9]
```diff
@@ -384,90 +378,100 @@ func isStringMapEqual(metadataA, metadataB map[string]string) bool {
...
 	glog.V(5).Infof("start to isVolumePublished, canAtMultiNode = %v, attachReq = %v",
 		canAtMultiNode, attachReq)
 
 	attachments, err := client.ListVolumeAttachments()
 	if err != nil {
-		glog.V(5).Info("ListVolumeAttachments failed: " + err.Error())
-		return nil, status.Error(codes.FailedPrecondition, err.Error())
+		msg := fmt.Sprintf("listVolumeAttachments failed: %v", err)
+		glog.Error(msg)
+		return status.Error(codes.FailedPrecondition, msg)
```
[the remark this answers is not in the document]
Comment (jackhaibo) [reply csi/server/plugin/opensds/controller.go]: Done.

File: csi/server/plugin/opensds/controller.go [hunk 10]
```diff
@@ -384,90 +378,100 @@ func isStringMapEqual(metadataA, metadataB map[string]string) bool {
 }
 
 // isVolumePublished Check if the volume is published and compatible
-func isVolumePublished(canAtMultiNode bool, attachReq *model.VolumeAttachmentSpec,
-	metadata map[string]string) (*model.VolumeAttachmentSpec, error) {
+func isVolumeCanBePublished(canAtMultiNode bool, attachReq *model.VolumeAttachmentSpec,
+	volMultiAttach bool) error {
+
 	glog.V(5).Infof("start to isVolumePublished, canAtMultiNode = %v, attachReq = %v",
 		canAtMultiNode, attachReq)
```
[the remark this answers is not in the document]
Comment (jackhaibo) [reply csi/server/plugin/opensds/controller.go]: Done.

**Pull request merged**